File: unit-tests/dispatcher.php
<?php

/*
  +------------------------------------------------------------------------+
  | Phalcon Framework                                                      |
  +------------------------------------------------------------------------+
  | Copyright (c) 2011-2012 Phalcon Team (http://www.phalconphp.com)       |
  +------------------------------------------------------------------------+
  | This source file is subject to the New BSD License that is bundled     |
  | with this package in the file docs/LICENSE.txt.                        |
  |                                                                        |
  | If you did not receive a copy of the license and are unable to         |
  | obtain it through the world-wide-web, please send an email             |
  | to license@phalconphp.com so we can send you a copy immediately.       |
  +------------------------------------------------------------------------+
  | Authors: Andres Gutierrez <andres@phalconphp.com>                      |
  |          Eduar Carvajal <eduar@phalconphp.com>                         |
  +------------------------------------------------------------------------+
*/

error_reporting(E_ALL | E_NOTICE | E_STRICT);

//Required for run in GDB
if(PHP_OS=='Linux'){
	if($_SERVER['USERNAME']=='gutierrezandresfelipe'){
		chdir('/home/gutierrezandresfelipe/phalcon');
	}
}

$dispatcher = new Phalcon_Dispatcher();

$request = Phalcon_Request::getInstance();
$response = Phalcon_Response::getInstance();

$basePath = './';
$dispatcher->setBasePath($basePath);
if($dispatcher->getBasePath()==$basePath){
	echo 'Phalcon_Dispatcher::setBasePath() [OK]', PHP_EOL;
} else {
	echo 'Phalcon_Dispatcher::setBasePath() [FAILED]', PHP_EOL;
	return;
}

$controllersDir = 'tests/controllers/';
$dispatcher->setControllersDir($controllersDir);
if($dispatcher->getControllersDir()==$controllersDir){
	echo 'Phalcon_Dispatcher::setControllersDir() [OK]', PHP_EOL;
} else {
	echo 'Phalcon_Dispatcher::setControllersDir() [FAILED]', PHP_EOL;
	return;
}

$dispatcher->setControllerName('index');
$dispatcher->setActionName('index');
$dispatcher->setParams(array());

try {
	$dispatcher->dispatch($request, $response);
}
catch(Phalcon_Exception $e){
	if($e->getMessage()=="File for controller class IndexController doesn't exists"){
		echo 'Phalcon_Dispatcher::dispatch() [1] [OK]', PHP_EOL;
	} else {
		echo 'Phalcon_Dispatcher::dispatch() [1] [FAILED]', PHP_EOL;
		return;
	}
}

$dispatcher->setControllerName('essai');
$dispatcher->setActionName('index');
$dispatcher->setParams(array());

try {
	$dispatcher->dispatch($request, $response);
}
catch(Phalcon_Exception $e){
	if($e->getMessage()=="File for controller class EssaiController doesn't exists"){
		echo 'Phalcon_Dispatcher::dispatch() [2] [OK]', PHP_EOL;
	} else {
		echo 'Phalcon_Dispatcher::dispatch() [2] [FAILED]', PHP_EOL;
		return;
	}
}

$dispatcher->setControllerName('test0');
$dispatcher->setActionName('index');
$dispatcher->setParams(array());

try {
	$dispatcher->dispatch($request, $response);
}
catch(Phalcon_Exception $e){
	if($e->getMessage()=="Class Test0Controller was not found on controller file"){
		echo 'Phalcon_Dispatcher::dispatch() [3] [OK]', PHP_EOL;
	} else {
		echo 'Phalcon_Dispatcher::dispatch() [3] [FAILED]', PHP_EOL;
		return;
	}
}

$dispatcher->setControllerName('test1');
$dispatcher->setActionName('index');
$dispatcher->setParams(array());

try {
	$dispatcher->dispatch($request, $response);
}
catch(Phalcon_Exception $e){
	if($e->getMessage()=="Action 'index' was not found on controller 'test1'"){
		echo 'Phalcon_Dispatcher::dispatch() [4] [OK]', PHP_EOL;
	} else {
		echo 'Phalcon_Dispatcher::dispatch() [4] [FAILED]', PHP_EOL;
		return;
	}
}

$dispatcher->setControllerName('test2');
$dispatcher->setActionName('index');
$dispatcher->setParams(array());
$controller = $dispatcher->dispatch($request, $response);
if(get_class($controller)=='Test2Controller'){
	echo 'Phalcon_Dispatcher::dispatch() [5] [OK]', PHP_EOL;
} else {
	echo 'Phalcon_Dispatcher::dispatch() [5] [FAILED]', PHP_EOL;
	return;
}

$dispatcher->setControllerName('test2');
$dispatcher->setActionName('essai');
$dispatcher->setParams(array());

try {
	$dispatcher->dispatch($request, $response);
}
catch(Phalcon_Exception $e){
	if($e->getMessage()=="Action 'essai' was not found on controller 'test2'"){
		echo 'Phalcon_Dispatcher::dispatch() [6] [OK]', PHP_EOL;
	} else {
		echo 'Phalcon_Dispatcher::dispatch() [6] [FAILED]', PHP_EOL;
		return;
	}
}

$dispatcher->setControllerName('test2');
$dispatcher->setActionName('other');
$dispatcher->setParams(array());
$controller = $dispatcher->dispatch($request, $response);
if(get_class($controller)=='Test2Controller'){
	echo 'Phalcon_Dispatcher::dispatch() [7] [OK]', PHP_EOL;
} else {
	echo 'Phalcon_Dispatcher::dispatch() [7] [FAILED]', PHP_EOL;
	return;
}

$dispatcher->setControllerName('test2');
$dispatcher->setActionName('another');
$dispatcher->setParams(array());
$dispatcher->dispatch($request, $response);
$value = $dispatcher->getReturnedValue();
if($value==100){
	echo 'Phalcon_Dispatcher::getReturnedValue() [1] [OK]', PHP_EOL;
} else {
	echo 'Phalcon_Dispatcher::getReturnedValue() [1] [FAILED]', PHP_EOL;
	return;
}

$dispatcher->setControllerName('test2');
$dispatcher->setActionName('anotherTwo');
$dispatcher->setParams(array(2, "3"));
$dispatcher->dispatch($request, $response);
$value = $dispatcher->getReturnedValue();
if($value==5){
	echo 'Phalcon_Dispatcher::getReturnedValue() [2] [OK]', PHP_EOL;
} else {
	echo 'Phalcon_Dispatcher::getReturnedValue() [2] [FAILED]', PHP_EOL;
	return;
}

$dispatcher->setControllerName('test2');
$dispatcher->setActionName('anotherthree');
$dispatcher->setParams(array());
$dispatcher->dispatch($request, $response);
$value = $dispatcher->getActionName();
if($value=='anotherfour'){
	echo 'Phalcon_Dispatcher::getActionName() [2] [OK]', PHP_EOL;
} else {
	echo 'Phalcon_Dispatcher::getActionName() [2] [FAILED]', PHP_EOL;
	return;
}

File: ext/phalcon/mvc/dispatcher.c

#ifdef HAVE_CONFIG_H
#include "../../ext_config.h"
#endif

#include <php.h>
#include "../../php_ext.h"
#include "../../ext.h"

#include <Zend/zend_operators.h>
#include <Zend/zend_exceptions.h>
#include <Zend/zend_interfaces.h>

#include "kernel/main.h"
#include "kernel/object.h"
#include "kernel/operators.h"
#include "kernel/memory.h"
#include "kernel/exception.h"
#include "kernel/fcall.h"


/*
 +------------------------------------------------------------------------+
 | Phalcon Framework                                                      |
 +------------------------------------------------------------------------+
 | Copyright (c) 2011-2013 Phalcon Team (http://www.phalconphp.com)       |
 +------------------------------------------------------------------------+
 | This source file is subject to the New BSD License that is bundled     |
 | with this package in the file docs/LICENSE.txt.                        |
 |                                                                        |
 | If you did not receive a copy of the license and are unable to         |
 | obtain it through the world-wide-web, please send an email             |
 | to license@phalconphp.com so we can send you a copy immediately.       |
 +------------------------------------------------------------------------+
 | Authors: Andres Gutierrez <andres@phalconphp.com>                      |
 |          Eduar Carvajal <eduar@phalconphp.com>                         |
 +------------------------------------------------------------------------+
 */
/**
 * Phalcon\Mvc\Dispatcher
 *
 * Dispatching is the process of taking the request object, extracting the module name,
 * controller name, action name, and optional parameters contained in it, and then
 * instantiating a controller and calling an action of that controller.
 *
 *<code>
 *
 *	$di = new Phalcon\Di();
 *
 *	$dispatcher = new Phalcon\Mvc\Dispatcher();
 *
 *  $dispatcher->setDI($di);
 *
 *	$dispatcher->setControllerName('posts');
 *	$dispatcher->setActionName('index');
 *	$dispatcher->setParams(array());
 *
 *	$controller = $dispatcher->dispatch();
 *
 *</code>
 */
ZEPHIR_INIT_CLASS(Phalcon_Mvc_Dispatcher) {

	ZEPHIR_REGISTER_CLASS_EX(Phalcon\\Mvc, Dispatcher, phalcon, mvc_dispatcher, phalcon_dispatcher_ce, phalcon_mvc_dispatcher_method_entry, 0);

	zend_declare_property_string(phalcon_mvc_dispatcher_ce, SL("_handlerSuffix"), "Controller", ZEND_ACC_PROTECTED TSRMLS_CC);
	zend_declare_property_string(phalcon_mvc_dispatcher_ce, SL("_defaultHandler"), "index", ZEND_ACC_PROTECTED TSRMLS_CC);
	zend_declare_property_string(phalcon_mvc_dispatcher_ce, SL("_defaultAction"), "index", ZEND_ACC_PROTECTED TSRMLS_CC);

	zend_class_implements(phalcon_mvc_dispatcher_ce TSRMLS_CC, 1, phalcon_mvc_dispatcherinterface_ce);

	return SUCCESS;

}

/**
 * Sets the default controller suffix
 *
 * @param string controllerSuffix
 */
PHP_METHOD(Phalcon_Mvc_Dispatcher, setControllerSuffix) {

	zval *controllerSuffix_param = NULL;
	zval *controllerSuffix = NULL;

	ZEPHIR_MM_GROW();
	zephir_fetch_params(1, 1, 0, &controllerSuffix_param);

		if (Z_TYPE_P(controllerSuffix_param) != IS_STRING) {
				zephir_throw_exception_string(spl_ce_InvalidArgumentException, SL("Parameter 'controllerSuffix' must be a string") TSRMLS_CC);
		}

		controllerSuffix = controllerSuffix_param;


	zephir_update_property_this(this_ptr, SL("_handlerSuffix"), controllerSuffix TSRMLS_CC);
	ZEPHIR_MM_RESTORE();

}

/**
 * Sets the default controller name
 *
 * @param string controllerName
 */
PHP_METHOD(Phalcon_Mvc_Dispatcher, setDefaultController) {

	zval *controllerName_param = NULL;
	zval *controllerName = NULL;

	ZEPHIR_MM_GROW();
	zephir_fetch_params(1, 1, 0, &controllerName_param);

		if (Z_TYPE_P(controllerName_param) != IS_STRING) {
				zephir_throw_exception_string(spl_ce_InvalidArgumentException, SL("Parameter 'controllerName' must be a string") TSRMLS_CC);
		}

		controllerName = controllerName_param;


	zephir_update_property_this(this_ptr, SL("_defaultHandler"), controllerName TSRMLS_CC);
	ZEPHIR_MM_RESTORE();

}

/**
 * Sets the controller name to be dispatched
 *
 * @param string controllerName
 */
PHP_METHOD(Phalcon_Mvc_Dispatcher, setControllerName) {

	zval *controllerName_param = NULL;
	zval *controllerName = NULL;

	ZEPHIR_MM_GROW();
	zephir_fetch_params(1, 1, 0, &controllerName_param);

		if (Z_TYPE_P(controllerName_param) != IS_STRING) {
				zephir_throw_exception_string(spl_ce_InvalidArgumentException, SL("Parameter 'controllerName' must be a string") TSRMLS_CC);
		}

		controllerName = controllerName_param;


	zephir_update_property_this(this_ptr, SL("_handlerName"), controllerName TSRMLS_CC);
	ZEPHIR_MM_RESTORE();

}

/**
 * Gets last dispatched controller name
 *
 * @return string
 */
PHP_METHOD(Phalcon_Mvc_Dispatcher, getControllerName) {


	RETURN_MEMBER(this_ptr, "_handlerName");

}

/**
 * Throws an internal exception
 *
 * @param string message
 * @param int exceptionCode
 */
PHP_METHOD(Phalcon_Mvc_Dispatcher, _throwDispatchException) {

	int exceptionCode;
	zval *message_param = NULL, *exceptionCode_param = NULL, *eventsManager = NULL, *dependencyInjector, *response = NULL, *exception, *_0 = NULL, *_1 = NULL, *_2, *_3;
	zval *message = NULL;

	ZEPHIR_MM_GROW();
	zephir_fetch_params(1, 1, 1, &message_param, &exceptionCode_param);

		if (Z_TYPE_P(message_param) != IS_STRING) {
				zephir_throw_exception_string(spl_ce_InvalidArgumentException, SL("Parameter 'message' must be a string") TSRMLS_CC);
		}

		message = message_param;
	if (!exceptionCode_param) {
		exceptionCode = 0;	} else {
		exceptionCode = zephir_get_intval(exceptionCode_param);
	}


	dependencyInjector = zephir_fetch_nproperty_this(this_ptr, SL("_dependencyInjector"), PH_NOISY_CC);
	if ((Z_TYPE_P(dependencyInjector) != IS_OBJECT)) {
		ZEPHIR_INIT_VAR(_0);
		object_init_ex(_0, phalcon_mvc_dispatcher_exception_ce);
		ZEPHIR_INIT_VAR(_1);
		ZVAL_STRING(_1, "A dependency injection container is required to access the 'response' service", 1);
		ZEPHIR_INIT_VAR(_2);
		ZVAL_LONG(_2, 0);
		zephir_call_method_p2_noret(_0, "__construct", _1, _2);
		zephir_throw_exception(_0 TSRMLS_CC);
		ZEPHIR_MM_RESTORE();
		return;
	}
	ZEPHIR_INIT_NVAR(_0);
	ZEPHIR_INIT_NVAR(_1);
	ZVAL_STRING(_1, "response", 1);
	zephir_call_method_p1(_0, dependencyInjector, "getshared", _1);
	ZEPHIR_CPY_WRT(response, _0);
	ZEPHIR_INIT_NVAR(_0);
	ZVAL_LONG(_0, 404);
	ZEPHIR_INIT_NVAR(_1);
	ZVAL_STRING(_1, "Not Found", 1);
	zephir_call_method_p2_noret(response, "setstatuscode", _0, _1);
	ZEPHIR_INIT_VAR(exception);
	object_init_ex(exception, phalcon_mvc_dispatcher_exception_ce);
	ZEPHIR_INIT_NVAR(_0);
	ZVAL_LONG(_0, exceptionCode);
	zephir_call_method_p2_noret(exception, "__construct", message, _0);
	_3 = zephir_fetch_nproperty_this(this_ptr, SL("_eventsManager"), PH_NOISY_CC);
	ZEPHIR_CPY_WRT(eventsManager, _3);
	if ((Z_TYPE_P(eventsManager) == IS_OBJECT)) {
		ZEPHIR_INIT_NVAR(_0);
		ZEPHIR_INIT_NVAR(_1);
		ZVAL_STRING(_1, "dispatch:beforeException", 1);
		zephir_call_method_p3(_0, eventsManager, "fire", _1, this_ptr, exception);
		if (ZEPHIR_IS_FALSE(_0)) {
			RETURN_MM_BOOL(0);
		}
	}
	zephir_throw_exception(exception TSRMLS_CC);
	ZEPHIR_MM_RESTORE();
	return;

}

/**
 * Handles a user exception
 *
 * @param \Exception exception
 */
PHP_METHOD(Phalcon_Mvc_Dispatcher, _handleException) {

	zval *exception, *eventsManager, *_0, *_1;

	ZEPHIR_MM_GROW();
	zephir_fetch_params(1, 1, 0, &exception);



	eventsManager = zephir_fetch_nproperty_this(this_ptr, SL("_eventsManager"), PH_NOISY_CC);
	if ((Z_TYPE_P(eventsManager) == IS_OBJECT)) {
		ZEPHIR_INIT_VAR(_0);
		ZEPHIR_INIT_VAR(_1);
		ZVAL_STRING(_1, "dispatch:beforeException", 1);
		zephir_call_method_p3(_0, eventsManager, "fire", _1, this_ptr, exception);
		if (ZEPHIR_IS_FALSE(_0)) {
			RETURN_MM_BOOL(0);
		}
	}
	ZEPHIR_MM_RESTORE();

}

/**
 * Possible controller class name that will be located to dispatch the request
 *
 * @return string
 */
PHP_METHOD(Phalcon_Mvc_Dispatcher, getControllerClass) {

	ZEPHIR_MM_GROW();

	zephir_call_method(return_value, this_ptr, "gethandlername");
	RETURN_MM();

}

/**
 * Returns the lastest dispatched controller
 *
 * @return Phalcon\Mvc\ControllerInterface
 */
PHP_METHOD(Phalcon_Mvc_Dispatcher, getLastController) {


	RETURN_MEMBER(this_ptr, "_lastHandler");

}

/**
 * Returns the active controller in the dispatcher
 *
 * @return Phalcon\Mvc\ControllerInterface
 */
PHP_METHOD(Phalcon_Mvc_Dispatcher, getActiveController) {


	RETURN_MEMBER(this_ptr, "_activeHandler");

}

